python/lsst/pex/config/config.py

```diff
@@ -453,20 +466,30 @@ def _parseTypingArgs(
             _typ = ForwardRef(unpackedParams)
             # type ignore below because typeshed seems to be wrong. It
             # indicates there are only 2 args, as it was in python 3.8, but
-            # 3.9+ takes 3 args. Attempt in old style and new style to
-            # work with both.
-            try:
-                result = _typ._evaluate(globals(), locals(), set())  # type: ignore
-            except TypeError:
-                # python 3.8 path
-                result = _typ._evaluate(globals(), locals())
+            # 3.9+ takes 3 args.
+            result = _typ._evaluate(globals(), locals(), frozenset())
             if result is None:
                 raise ValueError("Could not deduce type from input")
             unpackedParams = cast(type, result)
-        if "dtype" in kwds and kwds["dtype"] != unpackedParams:
-            raise ValueError("Conflicting definition for dtype")
-        elif "dtype" not in kwds:
-            kwds = {**kwds, **{"dtype": unpackedParams}}
+        # I would have expected this to be `typing.Union`, not
+        # `types.UnionType` from the docs, but testing says it's the latter.
+        if get_origin(unpackedParams) is types.UnionType:  # noqa: E721
+            match get_args(unpackedParams):
+                case [types.NoneType, annotated_dtype]:
+                    pass
+                case [annotated_dtype, types.NoneType]:
+                    pass
+                case _:
+                    raise ValueError("Unsupported union type for Field.")
+            annotated_optional = True
+        else:
+            annotated_dtype = unpackedParams
+            annotated_optional = False
```

Comment:
Consider this:
```suggestion
        match (get_origin(unpackedParams) is types.UnionType, get_args(unpackedParams)):  #  noqa: E721
            case (True as annotated_optional, (types.NoneType, annotated_dtype)):
                pass
            case (True as annotated_optional, (annotated_dtype, types.NoneType)):
                pass
            case (types.UnionType, _):
                raise ValueError("Unsupported union type for Field.")
            case (annotated_optional, _):
                annotated_dtype = unpackedParams
```

Reply:
Also you could import UnionType to shorten line 476

Reply:
I'm fine if you want to leave it, it definitely is in the area of different people  will find different things readable. But I reserve the right to highlight this in the future when you become a match operator aficionado.

For completeness, this is saying something like match ( (true or false), args), in the case it is a union type, either of the first two cases will match. The third case statement I forgot to update from a previous version I was going to suggest, and it too would be true.

in the case of the last one, you are simply saying bind the match argument to a variable no matter what it is, and we know that True would have matched above so the only thing it could be is false, so annotated_optional gets set as False. It is basically just the catch all _ that takes advantage of an assigment at the same time.